lighthouse-core/scripts/pptr-run-devtools.js

```diff
@@ -0,0 +1,161 @@
+    let result;
+    try {
+      // eslint-disable-next-line prefer-rest-params
+      result = original.apply(this, arguments);
+    } finally {
+      receiver[methodName] = original;
+    }
+    // In case of exception the override won't be called.
+    try {
+      // eslint-disable-next-line prefer-rest-params
+      Array.prototype.push.call(arguments, result);
+      // eslint-disable-next-line prefer-rest-params
+      override.apply(this, arguments);
+    } catch (e) {
+      throw new Error('Exception in overriden method \'' + methodName + '\': ' + e);
+    }
+    return result;
+  };
+}
+
+const sniffLhr = `
+new Promise(resolve => {
+  (${addSniffer.toString()})(
+    UI.panels.lighthouse.__proto__,
+    '_buildReportUI',
+    (lhr, artifacts) => resolve(lhr)
+  );
+});
+`;
+
+const startLighthouse = `
+(() => {
+  UI.ViewManager.instance().showView('lighthouse');
+  const button = UI.panels.lighthouse.contentElement.querySelector('button');
+  if (button.disabled) throw new Error('Start button disabled');
+  button.click();
+})()
+`;
+
+/**
+ * @param {import('puppeteer').Browser} browser
+ * @param {string} url
+ * @return {Promise<string>}
+ */
+async function testPage(browser, url) {
+  const page = await browser.newPage();
+
+  const targets = await browser.targets();
+  const inspectorTarget = targets.filter(t => t.url().includes('devtools'))[1];
+  if (!inspectorTarget) throw new Error('No inspector found.');
+
+  const session = await inspectorTarget.createCDPSession();
+  await session.send('Runtime.enable');
+
+  // Navigate to page and wait for initial HTML to be parsed before trying to start LH.
+  await new Promise(async resolve => {
+    const pageSession = await page.target().createCDPSession();
+    await pageSession.send('Page.enable');
+    pageSession.once('Page.domContentEventFired', resolve);
+    page.goto(url).catch(err => err);
+  });
+
+  /** @type {RuntimeEvaluateResponse|undefined} */
+  let startLHResponse;
+  while (!startLHResponse || startLHResponse.exceptionDetails) {
+    startLHResponse = await session.send('Runtime.evaluate', {expression: startLighthouse})
+      .catch(err => err);
+  }
+
+  /** @type {RuntimeEvaluateResponse} */
+  const remoteLhrResponse = await session.send('Runtime.evaluate', {
+    expression: sniffLhr,
+    awaitPromise: true,
+    returnByValue: true,
+  }).catch(err => err);
+
+  if (!remoteLhrResponse.result || !remoteLhrResponse.result.value) {
+    throw new Error('Problem sniffing LHR.');
+  }
+
+  await page.close();
+
+  return JSON.stringify(remoteLhrResponse.result.value);
+}
+
+/**
+ * @return {Promise<string[]>}
+ */
+async function readUrlList() {
+  /** @type {string[]} */
+  const urlList = [];
+
+  const rl = readline.createInterface(process.stdin, process.stdout);
+  rl.on('line', async line => {
+    if (line.startsWith('#')) return;
+    urlList.push(line);
+  });
+
+  return new Promise(resolve => rl.on('close', () => resolve(urlList)));
```

Comment:
FWIW with node 12, ReadableStreams are async iterables, so you can do 
```js
for await (const line of rl) {
  if (line.startsWith('#')) continue;
  urlList.push(line);
});

return urlList;
```
and it'll handle settling on `'close'` or any errors [automatically](https://262.ecma-international.org/11.0/#table-async-iterator-optional).

https://nodejs.org/api/readline.html#readline_example_read_file_stream_line_by_line